File: app/src/main/java/mitya/yahnc/network/RestService.java
package mitya.yahnc.network;

import mitya.yahnc.BuildConfig;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by Mitya on 02.07.2016.
 */
public abstract class RestService<T> {
    protected final T service;

    protected RestService(Class<T> serviceType) {
        this(serviceType, BuildConfig.API_URL);
    }

    protected RestService(Class<T> serviceType, String baseUrl) {
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(baseUrl)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .build();
        service = retrofit.create(serviceType);
    }

    public T getService() {
        return service;
    }
}

File: app/src/main/java/mitya/yahnc/network/RestService.kt
package mitya.yahnc.network

import mitya.yahnc.BuildConfig
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory

/**
 * Created by Mitya on 02.07.2016.
 */
abstract class RestService<T>(serviceType: Class<T>, baseUrl: String = BuildConfig.API_URL) {
    val service: T

    init {
        val loggingInterceptor = HttpLoggingInterceptor().apply { level = HttpLoggingInterceptor.Level.BODY }
        val okHttpClient = OkHttpClient.Builder()
                .addInterceptor(loggingInterceptor)
                .build()
        val retrofit = Retrofit.Builder()
                .baseUrl(baseUrl)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .client(okHttpClient)
                .build()
        service = retrofit.create(serviceType)
    }
}
